# Time series With LSTM

This is my first time series predictive ananlysis. The notebook uses sample dataset from an online sales company RST.

The challenge is to predict the sales of motorbikes, cars and minibuses

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
from statsmodels.tools.eval_measures import rmse
from sklearn.preprocessing import MinMaxScaler
from keras.preprocessing.sequence import TimeseriesGenerator
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
from keras.layers import Dropout
import datetime as dt
import warnings
warnings.filterwarnings("ignore")

Using TensorFlow backend.


In [2]:
data = pd.read_csv('D:\\R-Projects\\sales_data_sample.csv', encoding = 'unicode_escape')
data.head()

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

         ORDERDATE   STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0   2/24/2003 0:00  Shipped       1         2     2003  ...   
1    5/7/2003 0:00  Shipped       2         5     2003  ...   
2    7/1/2003 0:00  Shipped       3         7     2003  ...   
3   8/25/2003 0:00  Shipped       3         8     2003  ...   
4  10/10/2003 0:00  Shipped       4        10     2003  ...   

                    ADDRESSLINE1  ADDRESSLINE2           CITY STATE  \
0        897 Long Airport Avenue           NaN            NYC    NY   
1             59 rue de l'Abbaye           NaN          Reims   NaN   
2  27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3             78934 Hillside Dr.           NaN       Pasadena    CA   
4                7734 Strong St.           NaN  San Francisco    CA   

  POSTALCODE COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0      10022     USA       NaN              Yu             Kwai    Small  
1      51100  France      EMEA         Henriot             Paul    Small  
2      75508  France      EMEA        Da Cunha           Daniel   Medium  
3      90003     USA       NaN           Young            Julie   Medium  
4        NaN     USA       NaN           Brown            Julie   Medium  

[5 rows x 25 columns]

In [3]:
data.columns

Index(['ORDERNUMBER', 'QUANTITYORDERED', 'PRICEEACH', 'ORDERLINENUMBER',
       'SALES', 'ORDERDATE', 'STATUS', 'QTR_ID', 'MONTH_ID', 'YEAR_ID',
       'PRODUCTLINE', 'MSRP', 'PRODUCTCODE', 'CUSTOMERNAME', 'PHONE',
       'ADDRESSLINE1', 'ADDRESSLINE2', 'CITY', 'STATE', 'POSTALCODE',
       'COUNTRY', 'TERRITORY', 'CONTACTLASTNAME', 'CONTACTFIRSTNAME',
       'DEALSIZE'],
      dtype='object')

In [4]:
data.shape

(2823, 25)

In [5]:
data.describe()

ORDERNUMBER  QUANTITYORDERED    PRICEEACH  ORDERLINENUMBER  \
count   2823.000000      2823.000000  2823.000000      2823.000000   
mean   10258.725115        35.092809    83.658544         6.466171   
std       92.085478         9.741443    20.174277         4.225841   
min    10100.000000         6.000000    26.880000         1.000000   
25%    10180.000000        27.000000    68.860000         3.000000   
50%    10262.000000        35.000000    95.700000         6.000000   
75%    10333.500000        43.000000   100.000000         9.000000   
max    10425.000000        97.000000   100.000000        18.000000   

              SALES       QTR_ID     MONTH_ID     YEAR_ID         MSRP  
count   2823.000000  2823.000000  2823.000000  2823.00000  2823.000000  
mean    3553.889072     2.717676     7.092455  2003.81509   100.715551  
std     1841.865106     1.203878     3.656633     0.69967    40.187912  
min      482.130000     1.000000     1.000000  2003.00000    33.000000  
25%     2203.430000     2.000000     4.000000  2003.00000    68.000000  
50%     3184.800000     3.000000     8.000000  2004.00000    99.000000  
75%     4508.000000     4.000000    11.000000  2004.00000   124.000000  
max    14082.800000     4.000000    12.000000  2005.00000   214.000000

In [6]:
data.dtypes

ORDERNUMBER           int64
QUANTITYORDERED       int64
PRICEEACH           float64
ORDERLINENUMBER       int64
SALES               float64
ORDERDATE            object
STATUS               object
QTR_ID                int64
MONTH_ID              int64
YEAR_ID               int64
PRODUCTLINE          object
MSRP                  int64
PRODUCTCODE          object
CUSTOMERNAME         object
PHONE                object
ADDRESSLINE1         object
ADDRESSLINE2         object
CITY                 object
STATE                object
POSTALCODE           object
COUNTRY              object
TERRITORY            object
CONTACTLASTNAME      object
CONTACTFIRSTNAME     object
DEALSIZE             object
dtype: object

The above results shows that our dataset need some cleaning harmonization.

In [3]:
# Checking the Missing Values in the remaining columns in the dataset and the percentage to fill

missing_df =data.isnull().sum(axis=0).reset_index()
missing_df.columns = ['Column Name', 'Missing Values Count']
missing_df['Filling Factor (%)']=(data.shape[0]-missing_df['Missing Values Count'])/data.shape[0]*100
missing_df.sort_values('Filling Factor (%)').reset_index(drop = True)

NameError: name 'data' is not defined

In [8]:
#Are there null values? let's look for all columns with null values using this column
droping_list_all=[]
for j in range(0,7):
    if not data.iloc[:, j].notnull().all():
        droping_list_all.append(j)        
        #print(df.iloc[:,j].unique())
droping_list_all

[]

In [9]:
# And now we fill nulls with mode in any columns
for j in range(0,25):        
        data.iloc[:,j]=data.iloc[:,j].fillna(data.iloc[:,j].mode())

In [10]:
data.isnull()

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER  SALES  \
0           False            False      False            False  False   
1           False            False      False            False  False   
2           False            False      False            False  False   
3           False            False      False            False  False   
4           False            False      False            False  False   
...           ...              ...        ...              ...    ...   
2818        False            False      False            False  False   
2819        False            False      False            False  False   
2820        False            False      False            False  False   
2821        False            False      False            False  False   
2822        False            False      False            False  False   

      ORDERDATE  STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  ADDRESSLINE1  \
0         False   False   False     False    False  ...         False   
1         False   False   False     False    False  ...         False   
2         False   False   False     False    False  ...         False   
3         False   False   False     False    False  ...         False   
4         False   False   False     False    False  ...         False   
...         ...     ...     ...       ...      ...  ...           ...   
2818      False   False   False     False    False  ...         False   
2819      False   False   False     False    False  ...         False   
2820      False   False   False     False    False  ...         False   
2821      False   False   False     False    False  ...         False   
2822      False   False   False     False    False  ...         False   

      ADDRESSLINE2   CITY  STATE  POSTALCODE  COUNTRY  TERRITORY  \
0            False  False  False       False    False      False   
1             True  False   True       False    False      False   
2             True  False   True       False    False      False   
3             True  False  False       False    False       True   
4             True  False  False        True    False       True   
...            ...    ...    ...         ...      ...        ...   
2818          True  False   True       False    False      False   
2819          True  False   True       False    False      False   
2820          True  False   True       False    False      False   
2821          True  False   True       False    False      False   
2822          True  False  False       False    False       True   

      CONTACTLASTNAME  CONTACTFIRSTNAME  DEALSIZE  
0               False             False     False  
1               False             False     False  
2               False             False     False  
3               False             False     False  
4               False             False     False  
...               ...               ...       ...  
2818            False             False     False  
2819            False             False     False  
2820            False             False     False  
2821            False             False     False  
2822            False             False     False  

[2823 rows x 25 columns]

In [11]:
#Let's check now our data if is clean
missing_data = data.isnull().sum(axis=0).reset_index()
missing_data.columns = ['Column_Name', 'Missing_Count']
missing_data = missing_data[missing_data['Missing_Count']>0]
missing_data = missing_data.sort_values(by='Missing_Count')
missing_data

Column_Name  Missing_Count
19    POSTALCODE             76
21     TERRITORY           1073
18         STATE           1486
16  ADDRESSLINE2           2520

In [12]:
#Let's parse the date time
data['ORDERDATE']=pd.to_datetime(data['ORDERDATE'])
data.head()

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

   ORDERDATE   STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0 2003-02-24  Shipped       1         2     2003  ...   
1 2003-05-07  Shipped       2         5     2003  ...   
2 2003-07-01  Shipped       3         7     2003  ...   
3 2003-08-25  Shipped       3         8     2003  ...   
4 2003-10-10  Shipped       4        10     2003  ...   

                    ADDRESSLINE1  ADDRESSLINE2           CITY STATE  \
0        897 Long Airport Avenue       Level 3            NYC    NY   
1             59 rue de l'Abbaye           NaN          Reims   NaN   
2  27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3             78934 Hillside Dr.           NaN       Pasadena    CA   
4                7734 Strong St.           NaN  San Francisco    CA   

  POSTALCODE COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0      10022     USA      EMEA              Yu             Kwai    Small  
1      51100  France      EMEA         Henriot             Paul    Small  
2      75508  France      EMEA        Da Cunha           Daniel   Medium  
3      90003     USA       NaN           Young            Julie   Medium  
4        NaN     USA       NaN           Brown            Julie   Medium  

[5 rows x 25 columns]

Let's do Segmentation with number of clusters chosen randomly

In [13]:
data['ORDERDATE'] = [d.date() for d in data['ORDERDATE']]
data.head()

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

    ORDERDATE   STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0  2003-02-24  Shipped       1         2     2003  ...   
1  2003-05-07  Shipped       2         5     2003  ...   
2  2003-07-01  Shipped       3         7     2003  ...   
3  2003-08-25  Shipped       3         8     2003  ...   
4  2003-10-10  Shipped       4        10     2003  ...   

                    ADDRESSLINE1  ADDRESSLINE2           CITY STATE  \
0        897 Long Airport Avenue       Level 3            NYC    NY   
1             59 rue de l'Abbaye           NaN          Reims   NaN   
2  27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3             78934 Hillside Dr.           NaN       Pasadena    CA   
4                7734 Strong St.           NaN  San Francisco    CA   

  POSTALCODE COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0      10022     USA      EMEA              Yu             Kwai    Small  
1      51100  France      EMEA         Henriot             Paul    Small  
2      75508  France      EMEA        Da Cunha           Daniel   Medium  
3      90003     USA       NaN           Young            Julie   Medium  
4        NaN     USA       NaN           Brown            Julie   Medium  

[5 rows x 25 columns]

Let's Calculate Recency, Frequency and Monetary value for each customer

Assuming that we are analyzing the next day of latest order date in the data set. Creating a variable 'snapshot date**' which is the latest date in data set.

Recency : Recency is the number of days between the customer's latest order date and the snapshot date
Frequency: Number of purchases made by the customer
MonetaryValue: Revenue generated by the customer

In [14]:
# Calculate Recency, Frequency and Monetary value for each customer

snapshot_date = data['ORDERDATE'].max() + dt.timedelta(days=1) #latest date in the data set
data_RFM = data.groupby(['CUSTOMERNAME']).agg({
    'ORDERDATE': lambda x: (snapshot_date - x.max()).days,
    'ORDERNUMBER': 'count',
    'SALES':'sum'})

#Renaming the columns
data_RFM.rename(columns={'ORDERDATE': 'Recency',
                   'ORDERNUMBER': 'Frequency',
                   'SALES': 'MonetaryValue'}, inplace=True)

In [15]:
data_RFM.head()

Recency  Frequency  MonetaryValue
CUSTOMERNAME                                              
AV Stores, Co.               196         51      157807.81
Alpha Cognac                  65         20       70488.44
Amica Models & Co.           265         26       94117.26
Anna's Decorations, Ltd       84         46      153996.13
Atelier graphique            188          7       24179.96

Let's categorize our customers into 4 segments(Randomly Chosen)

Recency/Frequency/MonetaryValue : Level 4 > Level 3> Level 2 > Level 1

Lower the recency, higher the Recency level
Higher the number of orders, higher the Frequency level
Higher the monetary value, higher the MonetaryValue level

In [16]:
#Dividing into segments

# Create a spend quartile with 4 groups - a range between 1 and 5
MonetaryValue_quartile = pd.qcut(data_RFM['MonetaryValue'], q=4, labels=range(1,5))
Recency_quartile = pd.qcut(data_RFM['Recency'], q=4, labels=list(range(4, 0, -1)))
Frequency_quartile = pd.qcut(data_RFM['Frequency'], q=4, labels=range(1,5))


# Assign the quartile values to the Spend_Quartile column in data
data_RFM['R'] = Recency_quartile
data_RFM['F'] = Frequency_quartile
data_RFM['M'] = MonetaryValue_quartile

#df_RFM[['MonetaryValue_Quartile','Recency_quartile','Frequency_quartile']] = [MonetaryValue_quartile,Recency_quartile,Frequency_quartile]

# Print data with sorted Spend values
#print(df_RFM.sort_values('MonetaryValue'))

data_RFM.head()

Recency  Frequency  MonetaryValue  R  F  M
CUSTOMERNAME                                                       
AV Stores, Co.               196         51      157807.81  2  4  4
Alpha Cognac                  65         20       70488.44  4  2  2
Amica Models & Co.           265         26       94117.26  1  2  3
Anna's Decorations, Ltd       84         46      153996.13  3  4  4
Atelier graphique            188          7       24179.96  2  1  1

In [17]:
#Calculating RFM Score

data_RFM['RFM_Score'] = data_RFM[['R','F','M']].sum(axis=1)
data_RFM.head()

Recency  Frequency  MonetaryValue  R  F  M  RFM_Score
CUSTOMERNAME                                                                  
AV Stores, Co.               196         51      157807.81  2  4  4       10.0
Alpha Cognac                  65         20       70488.44  4  2  2        8.0
Amica Models & Co.           265         26       94117.26  1  2  3        6.0
Anna's Decorations, Ltd       84         46      153996.13  3  4  4       11.0
Atelier graphique            188          7       24179.96  2  1  1        4.0

Labelling the levels

RFM Score > 10 - High Value Customer

RFM SCore < 10 and RFM Score >= 6 - Mid Value Customer

RFM Score < 6 - Low Value Customer

In [18]:
#Naming Levels
# Define rfm_level function
def rfm_level(data):
    if np.bool(data['RFM_Score'] >= 10):
        return 'High Value Customer'
    elif np.bool((data['RFM_Score'] < 10) & (data['RFM_Score'] >= 6)):
        return 'Mid Value Customer'
    else:
        return 'Low Value Customer'

# Create a new variable RFM_Level
data_RFM['RFM_Level'] = data_RFM.apply(rfm_level, axis=1)

# Print the header with top 5 rows to the console
data_RFM.head()

Recency  Frequency  MonetaryValue  R  F  M  \
CUSTOMERNAME                                                          
AV Stores, Co.               196         51      157807.81  2  4  4   
Alpha Cognac                  65         20       70488.44  4  2  2   
Amica Models & Co.           265         26       94117.26  1  2  3   
Anna's Decorations, Ltd       84         46      153996.13  3  4  4   
Atelier graphique            188          7       24179.96  2  1  1   

                         RFM_Score            RFM_Level  
CUSTOMERNAME                                             
AV Stores, Co.                10.0  High Value Customer  
Alpha Cognac                   8.0   Mid Value Customer  
Amica Models & Co.             6.0   Mid Value Customer  
Anna's Decorations, Ltd       11.0  High Value Customer  
Atelier graphique              4.0   Low Value Customer

The above table shows AV is equal to 196

In [19]:
data.columns

Index(['ORDERNUMBER', 'QUANTITYORDERED', 'PRICEEACH', 'ORDERLINENUMBER',
       'SALES', 'ORDERDATE', 'STATUS', 'QTR_ID', 'MONTH_ID', 'YEAR_ID',
       'PRODUCTLINE', 'MSRP', 'PRODUCTCODE', 'CUSTOMERNAME', 'PHONE',
       'ADDRESSLINE1', 'ADDRESSLINE2', 'CITY', 'STATE', 'POSTALCODE',
       'COUNTRY', 'TERRITORY', 'CONTACTLASTNAME', 'CONTACTFIRSTNAME',
       'DEALSIZE'],
      dtype='object')

In [20]:
#Average Sales for all stores for Week

df_average_sales_week = final.groupby(by=['Date'], as_index=False)['Weekly_Sales'].sum()
df_average_sales = df_average_sales_week.sort_values('Weekly_Sales', ascending=False)

print(df_average_sales[:10])

#Seasonality vs Trend Analysis
plt.figure(figsize=(15,6))
plt.plot(df_average_sales_week.Date, df_average_sales_week.Weekly_Sales)
plt.show()

NameError: name 'final' is not defined

In [ ]:
#Sales variation during Holidays(Store/Dept)
holiday =  final[['Date', 'IsHoliday', 'Weekly_Sales']].copy()
holiday =  holiday.groupby(by=['Date','IsHoliday'], as_index=False)['Weekly_Sales'].sum()
holiday_group =  holiday.groupby(by=['IsHoliday'], as_index=False)['Weekly_Sales'].sum()
print( holiday_group)
#print( holiday[:5])

def holiday_sales(df):
    from matplotlib import pyplot as plt
    plt.figure(figsize=(15,6))
    labels = ['Date', 'IsHoliday_x', 'Weekly_Sales']
    plt.title('Sales Variation During Holidays')
    plt.plot(df.Date, df.Weekly_Sales)
    plt.show()
    
holiday_sales(holiday)

In [ ]:
# normalize the dataset and standadize it

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler(feature_range = (0, 1))

data = scaler.fit_transform(data)

In [ ]:
dataset.columns 